Add webkitgtk6 option

Sammy-T opened the pull request:
<!--

*********************************************************************
*               PLEASE READ BEFORE SUBMITTING YOUR PR               *
*     YOUR PR MAY BE REJECTED IF IT DOES NOT FOLLOW THESE STEPS     *
*********************************************************************

- *DO NOT* submit PRs for v3 alpha enhancements, unless you have opened a post on the discord channel.
  All enhancements must be discussed first.
  The feedback guide for v3 is here: https://v3alpha.wails.io/getting-started/feedback/

- Before submitting your PR, please ensure you have created and linked the PR to an issue.
- If a relevant issue already exists, please reference it in your PR by including `Fixes #<issue number>` in your PR description.
- Please fill in the checklists.

-->

# Description

This adds an option to run or build an app on Linux using webkitgtk-6.0.

Depends on:

- gtk4

Fixes #3193

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

> [!IMPORTANT]
> Features which have been removed/deprecated from webkit or gtk4 have been marked with todos.
> If they can be reimplemented, using this library option could potentially match the functionality of the current implementation.

# How Has This Been Tested?

- [ ] Windows
- [ ] macOS
- [x] Linux
      
Ubuntu 22.04

The performance is usually noticeable on an app with scrolled content or a complex layout.

Running/Building the app with this PR's changes requires:

- Setting the app's Linux [program name](https://wails.io/docs/reference/options#programname) option
- Adding `-tags webkit_6` to the Wails `dev` or `build` command

Warnings:

| Message | Reason? |
|---|---|
| `New application windows must be added after the GApplication::startup signal has been emitted.` | It should already be added after the startup signal so I think this warning is more about how it's set up. |
  
## Test Configuration

```bash
# Wails
Version         | v2.10.2
Package Manager | apt


# System
WARNING: failed to read int from file: open /sys/devices/system/cpu/cpu0/online: no such file or directory
┌──────────────────────────────────────────────────────────┐
| OS           | Ubuntu                                    |
| Version      | 22.04                                     |
| ID           | ubuntu                                    |
| Branding     |                                           |
| Go Version   | go1.24.5                                  |
| Platform     | linux                                     |
| Architecture | amd64                                     |
| CPU          | Intel(R) Core(TM) i5-6300HQ CPU @ 2.30GHz |
| Memory       | 16GB                                      |
└──────────────────────────────────────────────────────────┘

# Dependencies
┌──────────────────────────────────────────────────────────────────────────┐
| Dependency | Package Name          | Status    | Version                 |
| *docker    | docker.io             | Installed |                         |
| gcc        | build-essential       | Installed | 11.4.0                  |
| libgtk-3   | libgtk-3-dev          | Installed | 3.24.33-1ubuntu2.2      |
| libwebkit  | libwebkit2gtk-4.0-dev | Installed | 2.48.5-0ubuntu0.22.04.1 |
| npm        | npm                   | Installed | 11.5.2                  |
| *nsis      | nsis                  | Available | 3.08-2                  |
| pkg-config | pkg-config            | Installed | 0.29.2-1ubuntu3         |
|                                                                          |
└──────────────────────── * - Optional Dependency ─────────────────────────┘

# Diagnosis
Optional package(s) installation details:
  - nsis: sudo apt install nsis
```

**Requires webkitgtk-6 and gtk4.**

# Checklist:

- [ ] I have updated `website/src/pages/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Full WebKit6 Linux frontend: modern webview, GTK4-based window lifecycle and controls, menus, clipboard, keyboard accelerators, screen detection, file dialogs, inspector/DevTools, and asset serving.
  * WebKit6-backed request/response handling and clipboard APIs for the asset server and webview integration.

* **Chores**
  * Platform split and build-tag gating to select legacy WebKit or new WebKit6 paths at build time.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/wailsapp/wails/pull/4570?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Sammy-T committed:
Create webkit-6 versions of 'window' C files

Sammy-T committed:
Add webkit-6 to build constraints and cgo config

Sammy-T committed:
Construct gtk4 window

Sammy-T committed:
Manually create webview with content manager

`webkit_web_view_new_with_user_content_manager` has been removed in webkit-6 so this now has to be done manually.

Sammy-T committed:
Add webkit-6 alt key mask

Sammy-T committed:
Run and quit main loop

Sammy-T committed:
Add webkit-6 clipboard interaction

Sammy-T committed:
Connect buttons

Sammy-T committed:
Get gtk4 monitors

Sammy-T committed:
Implement menu and action handling

Gtk4 removed several menu widgets and reworked menu and accelerator interactions.

Sammy-T committed:
Create 'show-inspector' action

Sammy-T committed:
Clean up window and menu refs

Sammy-T committed:
Get and set window size

Sammy-T committed:
Clean up window_webkit6.c

This just removes some already disabled code and clarifies some comments.

Sammy-T committed:
Implement gtk4 message dialog

Sammy-T committed:
Implement gtk4 file chooser dialog

Sammy-T committed:
Enable webkit-6 gpu policy settings

Sammy-T committed:
Fix comment typo

Sammy-T committed:
Add gtk4 drag and drop foundation

This doesn't fully implement functionality and may fail.

See: https://discourse.gnome.org/t/file-drag-and-drop-in-gtkmm4/10548/5

Sammy-T committed:
Use program name as app id

Gtk4 requires an app id. The existing 'ProgramName' app option seems like a good fit for this.

Sammy-T committed:
Fix hardware acceleration selection

The 'on demand' policy was removed but 'never' should still map to case 2.

Sammy-T committed:
Merge branch 'master' into feature/webkitgtk6

COD3HUNT3R reviewed (commented):
Also need changes for doctor command

https://github.com/Sammy-T/wails/tree/feature/webkitgtk6/v2/internal/system/packagemanager

Sammy-T committed:
Remove `g_application_default_flags` redefinition

The redefinition might cause conflicts but it shouldn't be necessary.

(cherry picked from commit fcaddadfe4583152be228dd332d0022c1ba0f79b)

Sammy-T committed:
Switch to `GtkButtonsType`

(cherry picked from commit b0e94bef195011c7a27a8189aa5a104586795879)

Sammy-T commented:
Running this on X11 Linux Mint, this appears to be working nicely. 

I still don't know how/if I can get the todos (window positioning, etc.) working with webkit-6 though.
 
@leaanthony Are you able to test this and see how it runs for you? And do you know how to address the commented todos?

Sammy-T committed:
Use sync.map for action id to menu item mapping

Sammy-T commented on v2/internal/frontend/desktop/linux/menu_webkit6.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func processMenu(window *Window, menu *menu.Menu) {
+	gmenu := C.g_menu_new()
+	defer C.g_object_unref(C.gpointer(gmenu))
+
+	for _, menuItem := range menu.Items {
+		itemLabel := C.CString(menuItem.Label)
+		defer C.free(unsafe.Pointer(itemLabel))
+
+		submenu := processSubmenu(window, menuItem)
+		defer C.g_object_unref(C.gpointer(submenu))
+
+		C.g_menu_append_submenu(gmenu, itemLabel, C.toGMenuModel(submenu))
+	}
+
+	window.menubar = C.gtk_popover_menu_bar_new_from_model(C.toGMenuModel(gmenu))
+}
```
This should already be covered by the current defers.

Sammy-T committed:
Nil check main loop variable

Sammy-T committed:
Fix `sendShowInspectorMessage` definition

Sammy-T committed:
Remove `activate` from header

Sammy-T commented on v2/pkg/assetserver/webview/webkit6.go:
```diff
@@ -0,0 +1,5 @@
+//go:build linux && webkit_6
+
+package webview
+
+const Webkit2MinMinorVersion = 48
```
Webkit versioning is weird and doesn't match with the 6.0. See: https://webkitgtk.org/

Sammy-T committed:
Fix title setting

Sammy-T committed:
Check map result before asserting menu item

Sammy-T committed:
Revert extraneous c-string free

Applying the original suggestion didn't account for it being freed from C but at least there's a comment to help indicate it now.

Sammy-T committed:
Clean up previous menu

Sammy-T commented:
@leaanthony Thanks. It's been a lot of work and I wasn't sure I'd even get this far. Sorry if I've been spamming you. I wasn't expecting to get hit with all those suggestions. Some of them were relevant. I can switch it back to draft. 

Currently, this PR is working with the exception of missing some of the window control that was removed in gtk4 (I've marked them in the code as todo comments). 

I was hoping you could test it out and also take a look at the todos to see if you know of any workarounds/updates?

Sammy-T committed:
Fix webkit-6 window IsMinimized

Sammy-T committed:
Enable drag and drop

Sammy-T committed:
Set window transparency

This sets the transparency for the entire widget including the menu and the webview's contents.

Sammy-T committed:
Check for potential string concat overflow

Sammy-T committed:
Apply suggestions

Sammy-T committed:
Replace ulong with gulong

Sammy-T commented:
@leaanthony Are you able to look this over?

leaanthony reviewed (commented):
Quick question: just wondering why you went with #ifdef in some files and separating out the different versions using the filename convention in others. I'm not against it but the separation pattern is what we have elsewhere. Might be easier to read? Just curious

leaanthony commented:
I think it might be easier to follow/debug? Might be something an LLM can do reasonably quickly.

I'm going to try and get a v2 release out today and will look at including this in the next release. I'll install a fresh linux to test today.

Sammy-T committed:
Merge branch 'master' into feature/webkitgtk6

Sammy-T committed:
Replace webkit_6 ifdef with separate files

Sammy-T commented:
The `ifdef`s are replaced now.

leaanthony committed:
Merge branch 'master' into feature/webkitgtk6

Sammy-T commented on v2/internal/frontend/desktop/linux/keys_webkit6.go:
```diff
@@ -0,0 +1,107 @@
+	"f34":       C.guint(0xffdf),
+	"f35":       C.guint(0xffe0),
+	"numlock":   C.guint(0xff7f),
+}
```
I don't see any indication that these values have changed in gtk4.

Sammy-T committed:
Duplicate g_value to unset and free

Sammy-T committed:
Switch from stream_read_all to stream_read

leaanthony committed:
Merge branch 'master' into feature/webkitgtk6

leaanthony commented:
@Sammy-T - sorry for the delay in getting to this. Looking at it now along with a port to v3. @COD3HUNT3R - I'll sort out the v3 port, thanks for offering :+1:

leaanthony committed:
Merge branch 'master' into feature/webkitgtk6

leaanthony commented:
## Code review findings

Reviewed the full diff. Summary of issues found, roughly by severity:

### Blockers

| # | Issue | File |
|---|-------|------|
| 1 | **Format string vulnerability (CWE-134)** — `gtk_message_dialog_new`'s 5th arg is `options->message` directly; any `%s`/`%n` in the message body is interpreted as a format specifier. Fix: `"%s", options->message`. | `window_webkit6.c` |
| 2 | **Clipboard always empty for external content** — `gdk_clipboard_get_content()` only returns non-NULL when *your app* owns the clipboard. Any text copied from another app returns empty. Correct API: `gdk_clipboard_read_text_async` + `gdk_clipboard_read_text_finish` with a GMainContext spin loop. | `clipboard_webkit6.go` |
| 3 | **GObject leak on every monitor query** — `g_list_model_get_item` returns a ref the caller must unref. None of the screen/monitor code calls `g_object_unref` on returned `GdkMonitor*`. | `screen_webkit6.go`, `window_webkit6.c` |
| 4 | **`free` instead of `g_free` for accelerator string** — `gtk_accelerator_name` allocates with GLib's allocator; passing the result to stdlib `free` is undefined behaviour. Use `g_free`. | `menu_webkit6.go` |
| 5 | **Asset truncation via partial stream reads** — `g_input_stream_read` is documented to return fewer bytes than requested. No retry loop means large assets are silently truncated. Use `g_input_stream_read_all`. | `webkit6.go` |
| 6 | **`SoupMessageHeaders` reference leak** — `soup_message_headers_new` ref is never released after `webkit_uri_scheme_response_set_http_headers`. One ref leaked per asset request served. | `webkit6.go` |

### High priority

| # | Issue | File |
|---|-------|------|
| 7 | **NULL deref in drag-and-drop** — `g_file_get_path` returns NULL for non-local URIs (e.g. `smb://`, `http://`); `strlen(NULL)` follows. Guard with a NULL check. | `window_webkit6.c` |
| 8 | **Dual event loop** — `g_main_loop_run` is started in a goroutine alongside `g_application_run`. Running two GLib main loops simultaneously is unsupported and will produce unpredictable event dispatch. | `frontend_webkit6.go` |
| 9 | **`GDK_BACKEND=x11` forced unconditionally** — kills native Wayland support even though GTK4 supports it. Should only be set when `XDG_SESSION_TYPE` is `x11` or unset. | `frontend_webkit6.go` |
| 10 | **`defer` inside `for` loop** — all `C.CString` and `g_menu_new` allocations across loop iterations are held until function return, not freed at end of each iteration. | `menu_webkit6.go` |

### Silent no-ops (functional gaps)

These APIs are commented out with no implementation. Callers get no error and silently broken behaviour:

- `SetPosition` / `GetPosition` — always returns `(0, 0)` 
- `SetMinMaxSize` — does nothing
- `Center` — does nothing
- `SetKeepAbove` — does nothing
- `SetWindowIcon` — does nothing
- `Screen.IsPrimary` — always `false`

GTK4 did remove direct equivalents for some of these, but workarounds exist (X11 `_NET_WM_STATE_ABOVE` for always-on-top, monitor geometry for centering, signal-based size clamping for min/max). The v3 port has implementations for most of these.

### Low priority

- Debug `g_print` statements left in the `onDragDrop` callback (stdout noise in production)
- `strncat(paths, path, strlen(path))` — the safety argument is negated, this is identical to `strcat`
- `buttonPressedName` allocated and freed but never passed to anything
- `dialog.go` missing `!webkit_6` build tag (fragile, will break if GTK3-specific includes are added)
- `GtkFileChooserDialog` deprecated since GTK 4.10 (also present in v3; `GtkFileDialog` is the replacement but requires async handling)

---

Issues 3, 4, 5, 6 (GObject leak, wrong free, stream reads, soup leak) were also present in the v3 port and have been fixed in #5488. Issues 1, 7, 8, 9, 10 were either absent in v3 or already addressed differently.

Sammy-T committed:
Fix menu reference freeing

Sammy-T committed:
Unref monitors

Sammy-T committed:
Null check path

Sammy-T committed:
Use message with format

Sammy-T committed:
Switch to stream_read_all

Sammy-T commented:
I addressed most of the numbered items. 

- Item 2: Haven't looked at.
- Item 9: Am I missing something? I'm not seeing where the issue is.
- Item 8: It's been a while since I looked at this code. I don't remember why it's like that.
- Item 6: Crashes with a `signal arrived during cgo execution` error when I try to add the unref.

@leaanthony, I don't know when I'll be able to properly revisit this code. But, in the meantime, you're plenty welcome to make whatever edits you need.
